### PR TITLE
feat: add Google Docs, Google Slides and HubSpot MCP connectors

### PR DESCRIPTION
**Pull request by yiboyasss**
## Summary

Adds three built-in MCP connectors — Google Docs, Google Slides, and HubSpot CRM — so the new agent templates (SEO Brief Writer, Research & Enricher, Email Agent, Email Lead Response) can export documents/slides and read/write CRM records. Google Docs and Slides reuse the existing `google` OAuth provider; HubSpot adds a new `hubspot` provider.

## Connectors

| App ID | Provider | Scopes | MCP server |
|---|---|---|---|
| `google-docs` | google | `documents`, `drive.file` | `xagent.web.tools.mcp.google_docs` |
| `google-slides` | google | `presentations`, `drive.file` | `xagent.web.tools.mcp.google_slides` |
| `hubspot` | hubspot (new) | `crm.objects.contacts.read/write`, `crm.objects.companies.read/write`, `crm.objects.deals.read` | `xagent.web.tools.mcp.hubspot` |

All three are OAuth stdio connectors following the existing gmail/google-drive pattern (access token injected via `env_mapping`).

## Changes

- **MCP servers** (`src/xagent/web/tools/mcp/`): `google_docs.py` (create/read/append/replace/batch-update docs, Markdown→Docs), `google_slides.py` (create presentations, add/read slides), `hubspot.py` (search/create/update contacts & companies, read deals, read/create notes).
- **Registry** (`builtin_mcp_registry.py`): adds the three app rows plus the `hubspot` OAuth provider (auth/token URLs, client id/secret from env).
- **Migration** (`20260720_seed_docs_slides_hubspot.py`): idempotently seeds the provider + three apps; downgrade removes them and drops the `hubspot` provider only if no HubSpot apps remain.
- **Config** (`example.env`): documents `HUBSPOT_CLIENT_ID` / `HUBSPOT_CLIENT_SECRET` / `HUBSPOT_REDIRECT_URI`.

## Test plan

- `pytest tests/alembic/test_20260720_seed_docs_slides_hubspot.py tests/web/api/test_public_mcp_connector_visibility.py`
- Verified the three MCP server modules import and register their tools.
- Manually verified end-to-end: connected Google Docs via OAuth and created a document through the agent.

### CODE REVIEW

File: src/xagent/web/tools/mcp/hubspot.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+            body={"properties": _parse_properties(properties_json)},
+        )
+        return _success(company=company)
+    except Exception as e:
+        logger.error(f"Error updating company: {e}")
+        return _error(str(e))
+
+
+@mcp.tool()
+def hubspot_get_contact_deals(contact_id: str) -> str:
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/hubspot.py]: Both `hubspot_get_contact_deals` (here) and `hubspot_get_contact_notes` (~line 282) call the associations endpoint with no `limit`/`after` params, never read the response's `paging.next.after` cursor, and never surface a `total`/`truncated`/`has_more` flag in the `_success(...)` payload. For `get_contact_notes`, the user-facing `limit` (default 20, max 100) is applied as a client-side slice *after* a single unpaged fetch, so it does not fetch more to satisfy the requested count.

A contact with more deals/notes than the API's default association page size is silently truncated with no signal to the caller. Since this read path feeds an LLM agent's decisions about a contact ("does this contact have open deals", "what was discussed last"), silent incompleteness can make the agent confidently report wrong information.

Recommend paginating the associations fetch (follow `paging.next.after` until the requested count is satisfied) and/or returning a `has_more`/`truncated` flag so the caller knows the result is partial.

File: src/xagent/migrations/versions/20260720_seed_docs_slides_hubspot.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    if "oauth_providers" not in existing_tables:
+        return
+
+    if "public_mcp_apps" in existing_tables:
+        remaining_hubspot_apps = bind.execute(
+            sa.select(sa.func.count())
+            .select_from(PUBLIC_MCP_APPS_TABLE)
+            .where(PUBLIC_MCP_APPS_TABLE.c.provider_name == "hubspot")
+        ).scalar()
```
Comment (rogercloud) [inline src/xagent/migrations/versions/20260720_seed_docs_slides_hubspot.py:200]: Non-blocking edge case: `downgrade()` deletes the `hubspot` oauth_providers row based only on `remaining_hubspot_apps == 0`, without tracking whether *this* migration created it. `POST /admin/mcp/providers` (`src/xagent/web/api/admin_mcp.py:277-306`) lets an admin create a custom OAuth provider with any `provider_name`, including "hubspot", with no reservation on built-in names. If an admin pre-created a custom `provider_name="hubspot"` before upgrading, a later downgrade with no other hubspot apps remaining would delete that admin-created row even though this migration never created it.

Narrow scenario, but worth a one-line guard — e.g. only delete the provider row when it matches the exact seeded shape, or track provenance.

File: tests/alembic/test_20260720_seed_docs_slides_hubspot.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
+"""Tests for the Google Docs, Google Slides, and HubSpot CRM MCP connector seed migration."""
```
Comment (rogercloud) [inline tests/alembic/test_20260720_seed_docs_slides_hubspot.py:1]: Nit (optional): all 5 tests create the `oauth_providers`/`public_mcp_apps` tables with the full column set, so the migration's "table absent" branches (2 in upgrade, 2 in downgrade) and `_filter_row`'s actual column-dropping behavior are never hit — only its identity passthrough is exercised. Low risk (`_filter_row` is a 2-line dict comprehension and the untested branches fail safe by skipping the insert). Consider one test with a reduced-column schema to exercise real column filtering.

File: src/xagent/web/tools/mcp/google_docs.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+def get_docs_service() -> Any:
+    return build("docs", "v1", credentials=_get_credentials())
+
+
+def get_drive_service() -> Any:
+    return build("drive", "v3", credentials=_get_credentials())
+
+
+def _resolve_document_id(document_id: str) -> str:
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/google_docs.py:67]: shrink: `_resolve_document_id` here and `_resolve_presentation_id` in `google_slides.py:49` are duplicated verbatim (search → group(1) → strip fallback); only the compiled regex differs. Both files already import from the shared `mcp/utils.py`. Extract `resolve_id_from_url(value, pattern)` into `utils.py` and pass each file's regex in — behavior is identical since the regex objects stay local. Net ~-7 lines.

**Review by rogercloud — COMMENTED** [review]
## Summary

This PR adds three built-in MCP connectors — Google Docs, Google Slides, and HubSpot CRM — following the established gmail/google-drive OAuth-stdio connector pattern. It introduces three new stdio MCP server modules (`google_docs.py`, `google_slides.py`, `hubspot.py`), corresponding entries in `builtin_mcp_registry.py`, and an Alembic migration (`20260720_seed_docs_slides_hubspot.py`) that idempotently seeds the `hubspot` OAuth provider plus three `public_mcp_apps` rows, with a reversing downgrade. Two test files are added/modified. Total diff: 8 files, 1403 insertions, 1 deletion. No GitHub issue is linked to this PR.

## Approach

The design is sound. It faithfully follows the existing gmail/google-drive OAuth-stdio connector pattern with no new abstractions and no provider special-casing. The access token reaches the MCP stdio server via `launch_config.env_mapping`, identical to gmail/drive/calendar. HubSpot fits cleanly as a new OAuth provider: its token-info endpoint (which carries the token in the URL path) is handled by the pre-existing `{{access_token}}` placeholder substitution in `src/xagent/web/api/auth.py:1346`, so no HubSpot-specific branching was needed. Secrets are handled consistently with existing providers (env vars, empty-string fallback), and migration idempotency is confirmed by the passing tests.

Two design-level notes, both non-blocking:
- The migration seed-row duplication against `builtin_mcp_registry.py` is the repo's established "frozen snapshot" convention, guarded by `test_seed_rows_match_registry`. Acceptable as-is.
- The downgrade deletes the `hubspot` oauth_providers row based only on a "no remaining hubspot apps" check, without tracking provenance. See the inline finding on the downgrade for the narrow admin-created-provider edge case.

## Prior review status

The single prior gemini-code-assist comment — use `.scalar()` instead of `.scalar_one()` in the migration downgrade for broader SQLAlchemy compatibility — is **confirmed fixed**. `src/xagent/migrations/versions/20260720_seed_docs_slides_hubspot.py:200` now uses `.scalar()`, and no `.scalar_one()` remains in the file.

## Findings

### MAJOR

**`src/xagent/web/tools/mcp/hubspot.py` — silent truncation in `hubspot_get_contact_deals` (~249-278) and `hubspot_get_contact_notes` (~282-313).**
Both call the contact associations endpoint with no `limit`/`after` params, never read the response's `paging.next.after` cursor, and never surface a `total`/`truncated`/`has_more` flag in the `_success(...)` response. For `get_contact_notes`, the user-facing `limit` (default 20, max 100) is applied as a client-side slice *after* a single unpaged fetch — it does not fetch more to satisfy the requested count. A contact with more deals/notes than the API's default association page size is therefore silently truncated with no signal to the caller. Because this is a read path feeding an LLM agent's decisions about a CRM contact ("does this contact have open deals", "what was discussed last"), silent incompleteness can lead the agent to confidently report wrong information. Recommend paginating the associations fetch (follow `paging.next.after` until the requested count is satisfied) and/or returning a `has_more`/`truncated` flag so the caller knows the result is partial.

### MINOR

**`src/xagent/web/tools/mcp/hubspot.py:73-99` (`_request`) — no rate-limit/backoff handling. Non-blocking, follow-up.**
A `requests.HTTPError` is wrapped as `RuntimeError` with the response body appended (so it is not a raw crash), but there is no `Retry-After` awareness or retry on 429. This is a byte-for-byte repo-wide template shared by `onedrive.py`, `outlook.py`, `teams.py`, `meta_graph.py`, and `linkedin.py` — none of them retry either. Not a HubSpot-specific regression; better handled as a follow-up ticket to add shared retry/backoff across all requests-based MCP tools.

**`src/xagent/web/tools/mcp/google_docs.py:36-56` and `google_slides.py:25-46` — unreachable refresh-token credential branch. Non-blocking, systemic.**
Both `_get_credentials` / `get_slides_service` build a `GOOGLE_REFRESH_TOKEN`/`GOOGLE_CLIENT_ID`/`GOOGLE_CLIENT_SECRET` branch that is unreachable in practice: the `env_mapping` for google-docs (`builtin_mcp_registry.py:200`) and google-slides (line 219) only maps `GOOGLE_ACCESS_TOKEN`, so those other env vars are never populated in the MCP subprocess. This is copied verbatim from the pre-existing `gmail.py:23-44` pattern (also in `google_drive.py`, `calendar.py`). Not introduced or worsened by this PR — flagging only as a systemic cleanup candidate across the whole Google MCP family.

**Test coverage — no unit tests for ~893 lines of new Docs/Slides/HubSpot logic. Non-blocking.**
`google_docs.py`, `google_slides.py`, and `hubspot.py` have zero unit tests for non-trivial, mockable logic (Markdown→Docs conversion, `batchUpdate` construction, `_request` error wrapping, association-ID assembly in `hubspot_create_note`, and the pagination/slicing above). Repo precedent is genuinely split: linkedin/facebook/instagram/meta_graph MCP tools have `Mock`/`requests`-based tests, while gmail/google_drive/calendar do not, and there is no CI coverage gate. Since this matches the untested half of existing norms, it is minor — but at minimum, adding tests for `hubspot._request` error handling and `hubspot_create_note` association assembly (mirroring `tests/web/tools/test_linkedin_mcp.py` / `test_facebook_mcp.py`) would be worthwhile.

(Two further minor items — the migration downgrade provenance edge case and a migration-test schema gap — are filed as inline comments on their specific lines.)

## Simplification opportunities

**`google_docs.py:67-72` (`_resolve_document_id`) and `google_slides.py:49-54` (`_resolve_presentation_id`)** are duplicated verbatim (same search → `group(1)` → strip fallback); only the compiled regex differs. Both files already import from the shared `mcp/utils.py`, which is the established cross-file sharing pattern for all 11 MCP tool files in this directory. Extracting `resolve_id_from_url(value, pattern)` into `utils.py` (passing each file's regex in) preserves identical match/no-match/multi-match behavior. **Net: -7 lines possible.**

## Review coverage

All criteria were evaluated: design/approach, correctness, test coverage, documentation, and code quality. Injection risk and secrets-in-diff were both explicitly checked and are clean — no shell/SQL/command injection surface introduced, and no credentials or tokens are committed (secrets flow through env vars with empty-string fallback, consistent with existing providers).

File: src/xagent/web/tools/mcp/hubspot.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        if response_text:
+            message = f"{message} - {response_text}"
+        raise RuntimeError(message) from exc
+
+    if response.status_code == 204 or not response.content:
+        return {}
+    return response.json()
+
+
+def _list_association_ids(path: str, max_results: int) -> tuple[list[Any], bool]:
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/hubspot.py:104]: [Low, non-blocking] `_list_association_ids` has no zero-progress guard: if the API returned a page with empty `results` but a truthy `after` cursor, this loop would spin forever. Can't happen with well-behaved HubSpot responses, so it's theoretical — but a one-line break would harden it, e.g. `if not page.get("results"): return ids, bool(after)`.

File: tests/alembic/test_20260720_seed_docs_slides_hubspot.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            )
+        )
+        with patch.object(migration, "op", _operations(connection)):
+            migration.upgrade()
+            migration.downgrade()
+        assert not {"google-docs", "google-slides", "hubspot"} & _app_ids(connection)
+        assert "hubspot" in _provider_names(connection)
+
+
+def test_upgrade_and_downgrade_with_reduced_column_schema(tmp_path):
```
Comment (rogercloud) [inline tests/alembic/test_20260720_seed_docs_slides_hubspot.py:224]: [Low, non-blocking] The docstring claims this test exercises the downgrade guard's column-absent fallback, but the reduced schema here still includes `name`/`auth_url`/`token_url`, so the guard's name-only fallback path remains untested. The `_filter_row` column-dropping coverage is genuine and separate — only the docstring's scope claim is inaccurate. Consider correcting the docstring (or adding a case that actually drops a guard column).

**Review by rogercloud — APPROVED** [review]
## Follow-up review (d8b97c9)

Thanks for the updates. This revision addresses all actionable findings from the previous review at 3ea4beb: HubSpot association listing now paginates properly, the migration downgrade gained a shape-matching provenance guard, `resolve_id_from_url` was extracted into a shared helper, and the migration/HubSpot test suites were meaningfully expanded. I re-verified each item against the current code and ran the affected tests — all pass.

### Status of prior findings

| # | Finding | Severity | Status |
|---|---------|----------|--------|
| 1 | HubSpot silent truncation (no pagination on contact deals/notes) | MAJOR | ✅ FIXED |
| 2 | Migration downgrade could delete an admin-created "hubspot" provider | MINOR (non-blocking) | ✅ FIXED |
| 3 | `resolve_id_from_url` duplicated across google_docs/google_slides | MINOR | ✅ FIXED |
| 4 | Migration test coverage gaps (table-absent branches, real `_filter_row` column-dropping) | MINOR (nit) | ✅ FIXED (see note) |
| 5 | HubSpot `_request` has no rate-limit/429 backoff | MINOR (non-blocking, repo-wide) | ⚪ NOT FIXED — out of scope by design |
| 6 | Dead refresh-token credential branch in google_docs/google_slides | MINOR (non-blocking, systemic) | ⚪ NOT FIXED — out of scope by design |

Notes:
- **#1**: The new `_list_association_ids` helper follows `paging.next.after`, caps at `max_results`, and returns an accurate `has_more` flag. Page-boundary, empty-first-page, and multi-page cases are all correct, and the 8 new tests in `tests/web/tools/test_hubspot_mcp.py` include a genuine 2-page pagination test asserting the request params.
- **#2**: Downgrade now compares `name`/`auth_url`/`token_url` before deleting the provider row (client_id/secret excluded as env-dependent), applying each comparison only when the column exists. `test_downgrade_preserves_admin_created_hubspot_provider` covers a divergent-shape admin provider.
- **#4**: `_filter_row` column-dropping and all four table-absent branches are now genuinely exercised.
- **#5 / #6**: Both were explicitly flagged non-blocking / out-of-scope in the original review (a repo-wide template gap and a pre-existing copied pattern respectively), so their absence here is expected — no action needed in this PR.

### New low-severity findings (non-blocking)

- **[Low] `_list_association_ids` has no zero-progress guard** — `src/xagent/web/tools/mcp/hubspot.py`. If the API ever returned a page with empty `results` but a truthy `after` cursor, the pagination loop would spin forever. This can't happen with well-behaved HubSpot responses, so it's theoretical — but a one-line defensive break would harden it.
- **[Low] Test docstring overclaims coverage** — `tests/alembic/test_20260720_seed_docs_slides_hubspot.py`. `test_upgrade_and_downgrade_with_reduced_column_schema`'s docstring says it exercises the downgrade guard's column-absent fallback, but the reduced schema still includes `name`/`auth_url`/`token_url`, so that fallback (name-only matching) remains untested. The `_filter_row` coverage in that test is real and separate; only the docstring's scope claim is inaccurate.

### Known limitation (accepted, not a defect)

The downgrade provenance guard matches on static columns, so it cannot distinguish the seeded provider from a canonical-shape admin-created HubSpot provider (same name + real HubSpot OAuth URLs, differing only in credentials). A complete fix would require tracking provenance explicitly (e.g. a sentinel/marker column) rather than shape-matching. This is a narrow, acceptable edge case for a downgrade path — noting it as a known limitation rather than requesting further work.

### Verdict

**Good to merge.** The one MAJOR finding (silent truncation) is properly fixed with real, asserting test coverage, and all other prior findings are either fixed or were non-blocking/out-of-scope by the original review's own design. The two new items are low-severity and non-blocking. Nothing here should hold up the merge.
